torch/nested/nested.py

```diff
@@ -0,0 +1,198 @@
+            new_data.copy_(data_)
+            new_data = new_data.to(dtype=dtype, device=device)
+            new_data = new_data.requires_grad_(requires_grad)
+            if pin_memory:
+                new_data = new_data.pin_memory()
+            tensors.append(new_data)
+
+        return NestedTensor(tensors)
+
+def as_nestedtensor(data, dtype=None, device=None):
+    ret = NestedTensor(data)
+    if dtype is not None:
+        ret = ret.to(dtype)
+    if device is not None:
+        ret = ret.to(device)
+    return ret
+
+def _verify_tensors(tensors):
+    for tensor in tensors:
+        assert torch.is_tensor(tensor)
+    if len(tensors):
+        dim = tensors[0].dim()
+        layout = tensors[0].layout
+        device = tensors[0].device
+        dtype = tensors[0].dtype
+        requires_grad = tensors[0].requires_grad
+        is_pinned = tensors[0].is_pinned()
+        for tensor in tensors:
+            if not (dim == tensor.dim() and
+                    layout == tensor.layout and
+                    device == tensor.device and
+                    dtype == tensor.dtype and
+                    requires_grad == tensor.requires_grad and
+                    is_pinned == tensor.is_pinned()):
+                raise ValueError("Each passed Tensor "
+                                 "must match in dim, layout, "
+                                 "device, dtype and requires_grad")
+    else:
+        # Carrying around information as member variables vs.
+        # checking one entry of the owned Tensors is annoying
+        # and error-prone. Carrying around an is_empty attribute
+        # to hide the fact that we carry around a list with a
+        # single empty Tensor is also annoying and error-prone.
+        # Both are not worth it for a minor feature.
+        raise ValueError("We do not support empty lists for now.")
+
+class NestedTensor(object):
+    # The attributes must match across all constiuents
+    #
+    # The NestedTensor's attributes then become that of its
+    # constiuents.
+    #
+    # The passed lists of tensors must be non-empty for now.
+    #
+    # Attributes:
+    #     dim
+    #     layout
+    #     device
+    #     dtype
+    #     requires_grad
+    #     is_pinned
+    def __init__(self, tensors):
+        self._tensors = tensors
+        _verify_tensors(self._tensors)
+
+    @property
+    def grad(self):
+        grads = [t.grad for t in self._tensors]
+        if any(grad is None for grad in grads):
+            assert all(grad is None for grad in grads)
+            return None
+        else:
+            return NestedTensor(grads)
+
+    @property
+    def data(self):
+        return NestedTensor([t.data for t in self._tensors])
+
+    @property
+    def dim(self):
+        if DEBUG:
+            _verify_tensors(self._tensors)
+        return self._tensors[0].dim
+
+    @property
+    def shape(self):
+        raise NotImplementedError()
+
+    @property
+    def requires_grad(self):
```

Comment:
I think tensors support setting requires_grad. Not necessary for this iteration but something to consider.